src/qmapshack/gis/rte/router/CRouterRoutino.cpp

```diff
@@ -248,19 +245,54 @@ void CRouterRoutino::buildDatabaseList()
                 continue;
             }
 
+            // qDebug() << "buildDatabase Prefix" << prefix;
+
 #ifdef Q_OS_WIN
             Routino_Database * data = Routino_LoadDatabase(dir.absolutePath().toLocal8Bit(), prefix.toLocal8Bit());
 #else
             Routino_Database * data = Routino_LoadDatabase(dir.absolutePath().toUtf8(), prefix.toUtf8());
 #endif
-            if(data)
+            qDebug() << "Loaded Routino DB" << dir.absolutePath().toUtf8().data() << "  " << prefix.toUtf8().data();
+
+            if(data == nullptr)
+            {
+                QMessageBox::critical(this, "Routino ...", xlateRoutinoError(Routino_errno), QMessageBox::Abort);
+                continue;
+            }
+            /* determine the profile to use for each database*/
+            QVariantMap dmap;
+            dmap["db"] = QVariant ((qulonglong)data);
+
+            /* check possible profiles.xml locations and use the first available */
+            dmap["profilesPath"] = "";
+            QStringList profilesPaths = {
+                dir.filePath(prefix+"-profiles.xml"),
+                dir.filePath("profiles.xml"),
+                setup->routinoPath("profiles.xml").toUtf8()
+            };
+            for(const QString& profilePath : profilesPaths)
+            {
+                QFileInfo pinfo(profilePath);
+                if( pinfo.isReadable())
+                {
+                    dmap["profilesPath"] = pinfo.filePath();
+                    break;
+                }
+            }
+            if( dmap["profilesPath"].toString().isEmpty() )
             {
-                comboDatabase->addItem(prefix.replace("_", " "), quint64(data));
+                QMessageBox::critical(this, "Routino...",tr("Could not find a profiles XML file in expected folders. Routino Routing will not function"), QMessageBox::Ok);
+                return;
             }
             else
             {
-                QMessageBox::critical(this, "Routino...", xlateRoutinoError(Routino_errno), QMessageBox::Abort);
+                /* check if profile will load - will abort if not good */
+                loadProfiles(dmap["profilesPath"].toString());
```

Comment:
`loadProfiles()` might throw an exception that is not caught -> app crash -> user looses data.

Reply:
Understand that.  Will investigate a way to avoid an abort.

Reply:
I propose to check the profiles file during builddatabaselist and if an error occurs, do not add  that database to the list (so it is not accessable) with a warning to fix the xml file.  I have tried this  and if there are multiple databases the good ones are still available but the faulty one is not.  This avoids any segmentation due to that file.

Reply:
[Routino _002](https://user-images.githubusercontent.com/1490471/84631344-8f8e9a80-af30-11ea-8385-05de742998ba.png)

Here is my error message for a corrupted profiles.xml.  It does not abort, but the database is put not in the combo list. So it cant be used.

Reply:
I have now shifted things around.  loadProfiles is called in buildDatabaseList.  If an error occurs, then that database is not included in the databases able to be used.  A warning messagebox is emitted, indicating that the profiles file needs fixing and that the database is not available to use.  No aborts or throws are generated. loadProfiles is still called as required in the routing routines to use the correct profile parameters.